src/main/java/io/jawg/osmcontributor/ui/activities/PhotoActivity.java

```diff
@@ -217,9 +217,7 @@ protected void onCreate(Bundle savedInstanceState) {
         // Init parameters.
         latitude = getIntent().getDoubleExtra("latitude", 0);
         longitude = getIntent().getDoubleExtra("longitude", 0);
-        poiId = getIntent().getLongExtra("poiId", 0);
-        imageAdapter = new ImageAdapter(this, poiId);
-        gridPhotos.setAdapter(imageAdapter);
+        imageAdapters = new HashMap<Long, ImageAdapter>();
```

Comment:
The purpose of an adapter is to handle dynamic content for collections (and allow the views to recycle themselves instead of creating new ones).
Loading multiple adapters for each POI is probably not the best way to go.
Is there something I am not seeing there?

If my understanding is correct, you want to change the content of the gridView when a new POI is selected, while keeping in memory previously selected POI adapters.
This has a high cost in memory and is a potential source for leaks.

If you want to avoid making the image http requests multiple times, I would suggest to rather use an Image Disk Cache with your Image fetch library (Picasso in this project if I remember correctly).

Let me know if this was the expected behaviour or if I was missing something.

Cool docs on the adapter & view itself:
https://developer.android.com/guide/topics/ui/layout/gridview.html
http://www.piwai.info/android-adapter-good-practices